### PR TITLE
Add option to fetch tags even if fetch-depth > 0

### LINKED ISSUE #448: Allow fetch-depth to fetch tags within that depth
I'm using github actions to push docker images to dockerhub and require tags to make sure my artifacts version is correct. 

I'm using `fetch-depth: 500` to avoid really long fetch times with my repo (yes, I know this is a problem on it's own). 

I would expected the default behavior of the action is to checkout tags _within_ that 500 depth limit, not use `--no-tags`

```
 /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=500 origin +cd1e2e1b5d307b7f4a2c6e49113da14cfc7b1c36:refs/remotes/origin/master
```

is the command that gets run

https://github.com/bitcoin-s/bitcoin-s/runs/1947728258#step:2:40

### PR DESCRIPTION
**Pull request by RobertWieczoreck**
Hi there!

This PR adds an optional `fetch-tags` option that can be used to override the default behavior of adding the `--no-tags` arg to the generated `git fetch` command if `fetch-depth` is configured and not `0`.

This way, users can use fetch-depth while still getting the tags inside this range, which can e.g.  be important when using automatic GIT versioning plugins like [jgitver](https://github.com/jgitver/jgitver).

Tested using a workflow on a separate branch that verifies function as well as backwards compatibility: See [Run 5432963197](https://github.com/RobertWieczoreck/checkout/actions/runs/5432963197/jobs/9880354020).

All Checks are green.

Also see [PR 2](https://github.com/RobertWieczoreck/checkout/pull/2) in my fork (description is mostly copied from there).

Fixes #448 

**EDIT:** Adapted to the recent changes regarding sparse checkout and added Jest tests.

### CODE REVIEW

**Comment by RobertWieczoreck** [comment]
Hello @thboop,
I don't really know what flow to follow for contributions and didn't find any guidelines (and it't my first one on GitHub..).
Could you please have a look and tell me how to get this reviewed and merged "correctly"? :)
Thanks in advance

**Comment by miluoshi** [comment]
Any updates on this? It looks like there are merge conflicts in this PR.

**Comment by RobertWieczoreck** [comment]
> Any updates on this? It looks like there are merge conflicts in this PR.

The maintainers don't seem to notice this PR, as they need to approve the workflow runs (and maybe review it).
I don't know if I did something wrong as it's my first contribution on GitHub.

Regarding he merge conflicts: I just had to fetch the latest changes from main to my fork.
The updated fork PR is [#2](https://github.com/RobertWieczoreck/checkout/pull/2). Still all checks green.
Also testet again with a workflow in a separate branch: [4255310531](https://github.com/RobertWieczoreck/checkout/runs/4255310531?check_suite_focus=true)
Still works.

I have another contribution in mind regarding the new [GIT sparse index](https://github.blog/2021-11-10-make-your-monorepo-feel-small-with-gits-sparse-index/), which yields extreme performance gains when working with large monorepos.
I may just publish my fork at some point if no reaction occurs here - already using it internally right from the fork branch.

**Comment by miluoshi** [comment]
@RobertWieczoreck thanks for your contribution! Hopefully one of maintainers notices the PR.

> I have another contribution in mind regarding the new GIT sparse index, which yields extreme performance gains when working with large monorepos.

I am curious about your contribution that uses git sparse index. I am working with a monorepo that would need some performance improvements in checkout process. Can you share any details?

**Comment by bonomat** [comment]
It would be cool if we could get this feature in. In our build process we want to include the latest git tags but at the moment this github action does not allow us to fetch them.

**Comment by RobertWieczoreck** [comment]
@miluoshi 
> I am curious about your contribution that uses git sparse index. I am working with a monorepo that would need some performance improvements in checkout process. Can you share any details?

I didn't have time to analyse the changes required to this action completely, but it would be a bigger change than this PR, as right now even a regular sparse-checkout isn't possible. So we would need at least:
1. A new option called `sparse-checkout` (or `sparse-directories`?) with default `false` that...
2. Triggers the sparse-checkout init  (smth. like adding `--cone --sparse-index` to the `git init` command?) and ...
3. Accepts a list of directories to be used for a `git sparse-checkout set` command prior to checkout
4. Another option `use-sparse-index` (default `true`), as some users might want to disable it due to compatibility issues with external tools that read the index

Notes:
- The exact commands needed could depend on the version of GIT used (already detected)
- The sparse-index feature is available from GIT 2.34 only

I need to perform some local tests on how to best integrate the feature with the current commands the action issues without breaking other use-cases and the commands available in different GIT versions.

I hope to be able to start working on this before the end of the year.

@bonomat and others in need of this feature:
You can already use it directly from my repository if you want. The company I work for already does so.

```
- name: Checkout with fetch-depth=50 and fetch-tags=true
    uses: RobertWieczoreck/checkout@21bbf8c0e4ac6f18360981a5a025a60a36f753ef
    with:
      fetch-depth: 50
      fetch-tags: true
```

This way you'll get exactly the changes of this PR (more secure), but miss updates from this main repo.
If you want to use the latest version of the branch (I try to keep it updated to the main repository when it changes), you could use `RobertWieczoreck/checkout@add-fetch-tags-option`.

@thboop, @ericsciple:
I still hope you'll find some time to review this PR / approve the workflow runs, so I try it with another mention :)

**Comment by EwoutH** [comment]
Any updates on this PR? I would love to be able to fetch tags without fetching the full repo's history!

**Comment by anthony-steele-cko** [comment]
Any update on this? I would very much like this option to  fetch tags without fetching the full repo's history.

**Comment by tiagovrtr** [comment]
This would be a great addition. looking forward to it

**Comment by derN3rd** [comment]
Sad to see, that there are so many good PRs that are not used/merged by the Github staff...

**Comment by markstijnman-shell** [comment]
Hope this gets picked up soon as well, would solve at least 3 issues from the looks of it: #338, #448, and #701.

**Comment by RobertWieczoreck** [comment]
I updated the branch once again and still hope to get some feedback from the maintainers.

**Comment by dlwiii** [comment]
The [RobertWieczoreck](https://github.com/RobertWieczoreck) fix solves the issue for us.   But it sure would be nice to be able to get tags using the standard checkout

**Comment by prashant-shahi** [comment]
any update on this?

**Comment by RobertWieczoreck** [comment]
I rebased the branch to the lastest changes from `main` (once again..).

```
- name: Checkout with fetch-depth=50 and fetch-tags=true
    uses: RobertWieczoreck/checkout@add-fetch-tags-option
    with:
      fetch-depth: 50
      fetch-tags: true
```
At this point I would recommend to just use my fork branch in your workflows, as I have no hope left to get feedback from the maintainers.
If anyone needs tags (maybe based on specific tags from here), just PM me or request it here.
Also don't hesitate to post me a message if the branch lags some important changes from here, as I only check this every month or so.

**Review by nikimahurin-RL — APPROVED** [review]

**Comment by RobertWieczoreck** [comment]
Thanks a lot for your approval, @nikimahurin-RL :)

Could you also approve the workflow runs? This is required because this is my first public GitHub contribution :/

**Comment by keenan-v1** [comment]
Why isn't this merged yet?

**Comment by gbpoole** [comment]
+1

Would be great if this could be actioned soon.

**Comment by kduret** [comment]
Hi @vmjoseph @vanZeben @rentziass 

Sorry to ping you, but is is possible you review this pull request cause it seems many people want this enhancement :)

**Review by ventsyv — APPROVED** [review]
Looks good

**Comment by RobertWieczoreck** [comment]
> Looks good. Can someone please merge it?
@ventsyv: Thanks for your review - I updated (rebased) the fork branch again to get rid of merge conflicts.

But still someone (a maintainer) has to approve the workflow runs :(

**Comment by php-coder** [comment]
It would be great to see it merged!

@TingluoHuang could you, please, take a look? Or ping someone who can? Thanks!

**Comment by kduret** [comment]
This PR is definitely ignored by maintainers :disappointed:

**Comment by coolcalcacol** [comment]
@fhammerl can you take a look please?

**Comment by jmgilman** [comment]
This has 82 thumbs up reactions. Why is it being ignored?

**Comment by kduret** [comment]
@RobertWieczoreck try to close this PR and open a new one :smirk:
This will put it at the top of the listing
Maybe maintainers will take a look... :eyes:

**Comment by ljharb** [comment]
@kduret all that will do is add more notifications to the pile that maintainers are obviously ignoring. Try filing a support ticket or something maybe, but don't create github noise.

**Comment by kduret** [comment]
@ljharb already contacted the support... and this is the answer : 
> You could reach out to any of the contributors. They should be able to take a look.
> I can see there are a number of pull requests open and I'm sure there are reviewers going through them.

**Comment by ljharb** [comment]
Then the best bet is to wait, even if it takes a decade.

**Comment by luketomlinson** [comment]
Hi @RobertWieczoreck, thanks for this PR. I think this is a valuable change, but I would ask that you could add a test for this? I am thinking it might make sense to break out building the `args` array inside of the `fetch` method and test that the args array looks correct with and without your new option? Once we have that test, I think it is reasonable to get this merged.

**Comment by pawamoy** [comment]
I can try to write such a test if @RobertWieczoreck is not willing to or doesn't have the resources anymore.

**Comment by pawamoy** [comment]
By the way, do I understand this PR correctly if I say that the following will fetch only the last commit but also *all* tags:

```yaml
    - name: Checkout
      uses: actions/checkout@v3
      with:
        fetch-tags: true
```

Or does this change only applies when fetch-depth is greater than 0? :thinking: 

My use-case: I run a tool in CI that detects API breaking changes (Python). By default it compares HEAD to the latest tag. It means I currently have to set `fetch-depth: 0`, but I don't need the whole history, only the tags / tagged commits. Not even sure if what I am asking for is possible :thinking:

**Comment by RobertWieczoreck** [comment]
> Hi @RobertWieczoreck, thanks for this PR. I think this is a valuable change, but I would ask that you could add a test for this? I am thinking it might make sense to break out building the `args` array inside of the `fetch` method and test that the args array looks correct with and without your new option? Once we have that test, I think it is reasonable to get this merged.

@luketomlinson 
I'm not sure whether this is what you expected, but I added tests for all relevant combinations of the `fetchDepth` and my new `fetchTags` option to `__test__/git-command-manager.test.ts`.
Please point me to the right direction if you meant something else :)

Still, a maintainer needs to approve the workflow runs as this is my first (public) contribution.
Or can it also be merged without this approval?

**Comment by RobertWieczoreck** [comment]
@pawamoy

Your post seems to be a little inconsistent to me:
The default value of `fetchDepth`  (if omitted as in your example) is `1`, as indicated by your first sentence.
Your question afterwards implies that you assume it was `0`.
Also you are using `actions/checkout@v3`, which doesn't contain this PR and thus doesn't support `fetchTags` at all.
To test this PR you can use `RobertWieczoreck/checkout@add-fetch-tags-option`. 

That said, to the answers :)

> Do I understand this PR correctly if I say that the following will fetch the last commit but also _all_ tags:

No, `fetchTags: true` will only fetch the tags on the commits included in the `fetchDepth`. 
So in your example `fetchDetph` is `1` (the default), which leads to the latest commit being fetched.
If this one commit has a tag, it is also fetched thanks to `fetchTags: true`. Without `fetchTags` the argument `--no-tags` would be added to the `git fetch` command line to avoid fetching it.
So you get 0 to 1 tags.
If you choose `fetchDepth: 20` and `fetchTags: true` you would get 20 commits and their tags, so at max 20 tags.

> Or does this change only apply when fetch-depth is greater than 0? 🤔

Yes, `fetchTags: true` only has an effect if `fetchTags > 0`. That's because for `<= 0` all commits and all tags are fetched anyway.

> UPDATE: sorry for the noise on this PR, I was able to make it work with the following additional step: [..]

That's a valid workaround I guess. Just be aware that the checkout action already did a fetch before according to its options.

> My use-case: I run a tool in CI that detects API breaking changes (Python). By default it compares HEAD to the latest tag. It means I currently have to set `fetch-depth: 0`, but I don't need the whole history, only the tags / tagged commits. Not even sure if what I am asking for is possible thinking

It's an interesting use-case imho and I would gladly expand this PR in covering it as it fits in well semantically.
Right now, `fetchDepth: true` just leads to `--no-tags` **not** being added ti the command line.
To get all tags independent of the `fetchDepth` (if > 0), one could add `--tags` to the command line. [See the docs](https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt---tags) for details.

### Suggestion

We could change the possible values of `fetchTags` to an enum with values `none` (default), `for-commits` and `all`.
`none` is the former `false` and `for-commits` the former `true` behavior, while `all` would cover your use-case and fetch all tags by adding `--tags` to the command line.

A second expansion would be to also consider `fetchTags` if `fetchTags <= 0`.
Then the default value of `fetchTags` would be set dynamically to:
- `all` if `fetchDepth <= 0`
- `none` if `fetchDepth > 0`

..to keep the old default behavior.
This would allow folks to fetch **all commits without any tags** by using `fetchDepth: 0` and `fetchTags: none`, which could be another valid use-case. Of course, `all` and `for-commits` are equivalent if `fetchDepth <= 0`.

What do you guys think about that?

**Comment by vanZeben** [comment]
> Run 5432963197

Hey @RobertWieczoreck,
What @luketomlinson was referring to was reworking [this section of code](https://github.com/actions/checkout/pull/579/files#diff-955a6019350fe74f3c0be90efdaa701e9b569a7d03c86d9b97a60125dd4d3a19R246-R270) and pulling it out into it's own function that you call within `fetch(...)` and then writing some tests about the expected output (contextual to your proposed changes) of that newly written function instead of just testing against what options are passed to the `git` command itself. 

Also know we're having an internal dialog about this right now as well and some relevant technical debt and what our best case is with this, and given the "stale-ness" of this PR from our side of things, this is top of mind right now.

**Comment by RobertWieczoreck** [comment]
@vanZeben 
I think I got it now, but I still don't really see why such a refactoring is related to this PR.
The same was true e.g. in PR #1369, which changed the `fetch` function [here](https://github.com/actions/checkout/pull/1369/files#diff-955a6019350fe74f3c0be90efdaa701e9b569a7d03c86d9b97a60125dd4d3a19R241) in a similar manner.
Said PR was BTW added on June 4 and merged 5 days later (!), while this PR was ignored for nearly 2 years now and is much smaller.

Also, the `fetch` function does exactly two things: Building the args array and executing `git` with it asynchronously. if we refactor the array building to a new function, the `fetch` function will become a two-liner (which by itself can be a good thing of course).
But: In my test, the `execGit` function is mocked, which means it already covers the array building and nothing else - as if it would test your proposed `buildArgs` function.

If there is a **good reason** for your proposed refactoring I missed, just tell me :slightly_smiling_face: 
I don't have much time right now anyways, so we can discuss some more days :thinking:

**Review by luketomlinson — APPROVED** [review]
Hi @RobertWieczoreck, we've decided to accept this PR as is. We'll let it bake in the main branch for a little bit, before doing an official release. In the meantime, you can use `actions/checkout@main` to get the changes. 

Thank you for the PR!

**Comment by MrChocolatine** [comment]
2 years to get this merged. So reasonable.

**Comment by ventsyv** [comment]
@MrChocolatine Though I can't fathom why it took so long, I guess better late than never ...

**Comment by marc-hb** [comment]
> We could change the possible values of fetchTags to an enum with values none (default), for-commits and all.

I find it disappointing that this new user interface adds a **boolean** when the actual git fetch option has **3 different values**. I've seen even expert git users wrongly believe that the git fetch option is a boolean. This new Github option will no doubt reinforce this common misconception.

It's even more disappointing to find that this issue was discussed in the code review :-(

The middle-ground, default, "for-commits" value can be extremely useful when trying to optimize CI for repos with a large number of branches and tags while preserving `git describe`. I guess that's why it's git's default.

**Comment by marc-hb** [comment]
On a totally different topic, anyone interested in saving `git clone` time, resources and money should take a look at git's `--filter` option:
- https://github.com/actions/checkout/issues/1152

It's incredibly easy to use and offers speedups ranging between 2 and 5.

**Comment by lukasz-mitka** [comment]
There's a bug when `fetch-tags: true` and job is triggered by the tag: https://github.com/actions/checkout/issues/1467
@RobertWieczoreck fyi

**Comment by SergeyAmshaev** [comment]
0

**Pull request merged**
